yarn 2 compatibility

cspotcode opened the pull request:
@merceyz

agilgur5 commented on src/createJestConfig.ts:
```diff
@@ -1,15 +1,30 @@
 import { Config } from '@jest/types';
+import { createRequire } from 'module';
```
Per [the docs](https://nodejs.org/api/module.html#module_module_createrequire_filename) you linked to in https://github.com/formium/tsdx/issues/688#issuecomment-724892954, `createRequire` only works Node 12+. TSDX supports Node 10+, so this API cannot be used.

cspotcode replied:
Switched to `create-require` which works on node 10.

agilgur5 commented on src/createJestConfig.ts:
```diff
@@ -1,15 +1,30 @@
 import { Config } from '@jest/types';
+import { createRequire } from 'module';
+import * as Path from 'path';
```
```suggestion
import path from 'path';
```

That is how it is imported in several files, so please be consistent.

agilgur5 reviewed (changes requested):
I only took a quick glance through here, but there already seem to be several issues. CI is also failing as `bahmutov/npm-install` does not support Yarn 2: https://github.com/bahmutov/npm-install/issues/22

The Jest issue is not the only issue with Yarn 2 support, there should be several others. But the Jest issue will likely be solved in a better fashion with presets #634 (though that still needs testing). This approach seems  different from Jest's defaults, which could cause inconsistencies and easy API breakage (and may itself be breaking).

cspotcode commented:
This PR is marked draft because it is still a work in progress, so it's expected to be broken.  I created it so I could ping @merceyz who wanted to follow along.

cspotcode commented:
I got all the tests passing in yarn2.  PR is still marked "draft" because the yarn folks have sent a bugfix to jest that may allow me to simplify and speed up the changes I had to make here.

cspotcode commented on package.json:
```diff
@@ -18,9 +18,7 @@
-  "bin": {
-    "tsdx": "./dist/index.js"
-  },
+  "bin": "./dist/index.js",
```
Yeah, this change was a mistake that snuck into `git add`.  Reverted.

agilgur5 reviewed (changes requested):
#### Reply

> This PR is marked draft because it is still a work in progress, so it's expected to be broken. I created it so I could ping `@merceyz` who wanted to follow along.

So a "Draft PR" indicates that it is "ready for feedback" but not necessarily ready for merging. If you did not intend to receive feedback, then it sounds like just having a branch fits the stage better than a Draft PR.

I was getting notifications of every commit here, so I had to unsubscribe to this thread entirely. Please note that you'll need to ping me directly for me to be notified as a result of that.

### New Review

It looks like this is still in draft and there's still a number of remaining items left to-do, but I took a look at some of the code (not all of it; I haven't really read through some of the functions, shell scripts, or CI caching/steps etc) and left some comments. Some of the previous issues I mentioned in previous comments have not been resolved and there are some additional issues.

The scope of this PR, in particular, changing test infra and some test structure in non-trivial ways may have made sense for a pre-start discussion to reduce future iterations.

This will still need at least one more review and I'll have to thoroughly manually test it as well if all the issues are resolved.

#### Testing

It might be necessary to add some new tests or change something here, but I'm not sure. In particular,
- `rollup-plugin-pnp-resolve` wasn't added but is mentioned in https://github.com/formium/tsdx/issues/688#issuecomment-616978687 and https://github.com/formium/tsdx/issues/458#issuecomment-578407378
- https://github.com/formium/tsdx/issues/458#issue-555061704 mentions an issue with TS installation with `tsdx build` and Berry
- https://github.com/formium/tsdx/issues/688#issuecomment-681054218 says `build` does work (differing from the above two), but says `lint` needs some changes (not sure what specifically).

This seems to only directly resolve the `test` portion with the `createRequire` code but not the others, so maybe there's something missing here, unless Berry has had some changes that make the other changes no longer required.

cspotcode commented on test/e2e/fixtures/build-default/package.json:
```diff
@@ -3,5 +3,11 @@
-  "license": "MIT"
+  "license": "MIT",
+  "dependencies": {
+    "@types/lodash": "4.14.165",
+    "@types/node": "14.14.7",
+    "lodash": "4.17.20",
+    "regenerator-runtime": "0.13.7"
```
yarn 2 requires dependencies to be explicitly declared, so without the declarations here, they fail.  For example, `tsdx build` fails when it can't find `@types/node`, so that needs to be explicitly declared.

The hard pinning was for performance to make sure tests all use the same versions that were already pulled into local cache.  I was getting network failures from running multiple `yarn` processes in parallel, all trying to make many requests to npm.  It's possible the hard pinning can be removed since IIRC I started copying a seed `yarn.lock` into each stage directory before running `yarn`

cspotcode commented on test/utils/fixture.ts:
```diff
@@ -10,20 +21,73 @@ export function setupStageWithFixture(
+    );
+  }
   shell.cd(stagePath);
+  if (getPackageManager() === 'yarn2') {
+    // Setup stage directory as a project root for yarn2
+    fs.writeFileSync(
+      '.yarnrc.yml',
+      `yarnPath: ${path.join(
+        rootDir,
+        'test/yarn2/.yarn/releases/yarn-sources.cjs'
+      )}\ncacheFolder: ${rootDir}/test/yarn2/.yarn/cache`
+    );
+    // Copy seed lockfile with most resolutions already performed.  Makes tests faster.
+    shell.exec(`cp ${rootDir}/test/yarn2/yarn.lock ${stagePath}/`);
+  }
+  packageManagerInstall();
```
The tests are running `tsdx` within yarn pnp, which means each stage directory needs to have its dependencies resolved.  With running the stages in parallel, they each need to have their own PnP resolutions instead of sharing with the project root.  Another option could be creating all `stage-*` directories at once, declaring them as workspaces of the root, and running `yarn` once to resolve them all.

This approach, running `yarn` in each stage directory, is still pretty fast thanks to the seed `yarn.lock` and the pre-populated cache.  `yarn` essentially checks the `yarn.lock`, sees that it already knows the full dependency tree, sees that it already has all modules in cache, and can stop there.

cspotcode commented on test/utils/fixture.ts:
```diff
@@ -10,20 +21,73 @@ export function setupStageWithFixture(
+    const libTesterPath = require.resolve('./lib-tester.js');
+    const args: libTester.JsonArgs = { libPath: absLibPath, expression };
+    const encodedValue = shell.exec(
+      `${isYarn2 ? 'yarn ' : ''}node ${libTesterPath} ${Buffer.from(
```
IIRC I set it up so that jest is still running out of yarn 1 `node_modules` (less changes to the tests) and only `tsdx` and the build output of `tsdx` are executed with PnP.  `yarn node` is responsible for installing PnP into the node process by setting the right `NODE_OPTIONS` env var.

To support unit testing, we can switch to running everything under yarn 2.  I tried that earlier on in my experimenting, so I can switch back to it.

cspotcode commented on test/utils/fixture.ts:
```diff
@@ -10,20 +21,73 @@ export function setupStageWithFixture(
+    // Add tsdx dependency pointed to project directory
+    const pkg = JSON.parse(fs.readFileSync('package.json', 'utf8'));
+    pkg.dependencies = pkg.dependencies ?? {};
+    pkg.dependencies['tsdx'] = `portal:${rootDir}`;
```
Portal causes the PnP resolver to resolve all dependencies of the portal-ed directory just like any other dependency.  Without this, the code in `../dist` cannot load its dependencies via PnP.

cspotcode replied:
Resolving because newest version of this PR avoids this change.  It keeps everything as workspaces of the root package.json

merceyz commented:
> rollup-plugin-pnp-resolve wasn't added but is mentioned in #688 (comment) and #458 (comment)

`resolve`, which is used by `@rollup/plugin-node-resolve`, has PnP support so that rollup plugin isn't needed anymore.

cspotcode commented:
@agilgur5 thanks for taking a look, I'm working through the review feedback.  I've left comments on a few things, am pushing fixes for some, and for other things like jest presets, I'll need to do some research.

It has been helpful to let Github Actions run the tests while I work, which is why I've been pushing code here as I go.  Understanding that you've muted this thread, is it ok if I keep pushing code as I go?  I can ping you only once I've addressed all review feedback.

cspotcode commented:
I got rid of the npm misnomer and trimmed down the test matrix.

I've reverted the package.json changes.  I did some testing and they were only necessary when running jest under PnP.  In another branch I'm going to re-implement the changes to run jest under PnP, which should simplify the tests, eliminate some of the messier changes to test infra, and enable unit-testing.  I'll see if/how many dependency changes are necessary.

timini commented:
Great to see so much progress on this, if there is a list of follow up tasks that would be great to get some others to input on those.

I mentioned before I would like to sponsor such efforts and I'd like to honour that, please sign up for issuehunt!

cspotcode commented on .github/workflows/nodejs.yml:
```diff
@@ -28,13 +28,14 @@ jobs:
       matrix:
         node: ['10.x', '12.x', '14.x']
         os: [ubuntu-latest, windows-latest, macOS-latest]
+        package_manager: [npm, yarn2]
```
Resolving because the misnomer has been fixed and test matrix has been shrunk

cspotcode commented on package.json:
```diff
@@ -77,7 +79,7 @@
     "execa": "^4.0.3",
     "fs-extra": "^9.0.0",
-    "jest": "^25.3.0",
+    "jest": "25.5.1",
```
Resolving because I've fixed to not pin and explained the upgrade in PR description.

cspotcode commented on package.json:
```diff
@@ -94,11 +96,12 @@
     "shelljs": "^0.8.3",
     "tiny-glob": "^0.2.6",
-    "ts-jest": "^25.3.1",
+    "ts-jest": "^26.4.4",
```
Resolving as I've updated the PR to avoid upgrading to v26

cspotcode commented on test/utils/fixture.ts:
```diff
@@ -10,20 +21,73 @@ export function setupStageWithFixture(
+export function getPackageManager() {
+  if (process.env.TSDX_TEST_PACKAGE_MANAGER === 'yarn2') return 'yarn2';
+  return 'npm';
+}
```
Resolving because I've fixed the misnomer and simplified to use `process.versions.pnp`

cspotcode commented on test/utils/fixture.ts:
```diff
@@ -10,20 +21,73 @@ export function setupStageWithFixture(
+}
+
+export function packageManagerInstall() {
+  if (getPackageManager() === 'yarn2') {
```
Resolving as inverted, and also has been simplified.

cspotcode commented on package.json:
```diff
@@ -94,11 +96,12 @@
+    "typescript": "^3.8.3"
   },
   "devDependencies": {
+    "@jest/types": "^25.5.0",
```
Resolving as added to description.

cspotcode commented on package.json:
```diff
@@ -123,6 +126,7 @@
     "styled-components": "^5.0.1",
     "tiny-invariant": "^1.1.0",
     "tiny-warning": "^1.0.3",
+    "type-fest": "0.13.1",
```
Resolving as unpinned.

cspotcode commented:
@timini how does issuehunt work?  Do I need to sign up, or does the tsdx team need to sign up?

As far as getting other people involved with the remaining work, I added a checklist of tasks to the issue description, but at the moment it's a bit messy.  I am also iterating on this other branch and merging to this PR when they're ready. 
 Trying to cut down on spam over here.  https://github.com/cspotcode/tsdx/pull/4

Perhaps more useful than the checklist, I added instructions for installing and testing this version of tsdx today.  Yarn 2 users, please feel free to give it a shot on your projects and let me know what's still broken.

timini commented:
Tested creating a yarn 2 monorepo using workspaces. So far i have no issues, i had to setup a tsconfig.json at the workspace root which is extended by the workspace packages

https://github.com/timini/tsdx-yarn-2-monorepo

timini commented:
I have done very minimal project setup so far, but it looked promising, its based off https://github.com/cspotcode/tsdx#yarn2

I haven't tried with main line tsdx. 

Are there any tests i should make? Is it possible to add tests for this kind of monorepo setup?

timini commented:
OK i tested the same setup using main branch tsdx and encountered issues with yarn install. It seems those issues are resolved by https://github.com/cspotcode/tsdx#yarn2

cspotcode commented:
@timini the "changed I've had to make" section of the OP lists fixes I made for compatibility.  The yarn2 bugfix has since been merged and released, so shouldn't be necessary any more.  Are you using the latest stable yarn 2?
